src/devices/bus/s100/vectordualmode.cpp

```diff
@@ -0,0 +1,315 @@
+		attotime sector_time = flop->time_next_index() - machine().time() + (rot_time / 16) * m_sector;
+		if (sector_time > rot_time)
+			sector_time -= rot_time;
+		m_sector_timer->adjust(sector_time, SECTOR_START);
+	}
+}
+
+bool s100_vector_dualmode_device::get_next_bit(attotime &tm, const attotime &limit)
+{
+	int bit = m_pll.get_next_bit(tm, m_floppy[m_drive]->get_device(), limit);
+	if (bit < 0)
+		return false;
+	m_pending_byte <<= 1;
+	m_pending_byte |= bit;
+	m_pending_size++;
+	return true;
+}
+
+TIMER_CALLBACK_MEMBER(s100_vector_dualmode_device::sector_cb)
+{
+	switch (static_cast<sector_timer_state>(param)) {
+	case SECTOR_START:
+		if (m_read) {
+			m_pll.set_clock(half_bitcell_size);
+			m_pll.read_reset(machine().time());
+			attotime tm;
+			attotime limit = machine().time() + half_bitcell_size*512;
+			while (get_next_bit(tm, limit)) {} // init PLL
+			limit += half_bitcell_size*16*30;
+			while (get_next_bit(tm, limit) && m_pending_byte != 0x5554) {}
+			if (m_pending_byte == 0x5554) {
+				m_pending_size = 1;
+				m_byte_timer->adjust(tm - machine().time());
+			}
+		} else {
+			m_pending_size = 0;
+			m_byte_timer->adjust(attotime::zero);
+		}
+		m_sector_timer->adjust(attotime::from_msec(200)/16, SECTOR_END);
+		break;
+
+	case SECTOR_END:
+		m_byte_timer->enable(false);
+		if (m_read)
+			m_ram[274] = 0; // Ignore ECC
+		break;
+	}
+}
+
+TIMER_CALLBACK_MEMBER(s100_vector_dualmode_device::byte_cb)
+{
+	if (m_read) {
+		if (m_pending_size == 16) {
+			m_pending_size = 0;
+			m_ram[m_cmar++] = unmfm_byte(m_pending_byte);
+			m_cmar &= 0x1ff;
+		}
+		attotime tm;
+		while (m_pending_size != 16 && get_next_bit(tm, attotime::never)) {}
+		m_byte_timer->adjust(tm - machine().time());
+	} else {
+		if (m_pending_size == 16) {
+			attotime start_time = machine().time() - half_bitcell_size*m_pending_size;
+			attotime tm = start_time + attotime::from_usec(1);
+			attotime buf[8];
+			int pos = 0;
+			while (m_pending_size) {
+				if (m_pending_byte & (1 << --m_pending_size))
+					buf[pos++] = tm;
+				tm += half_bitcell_size;
+			}
+			floppy_image_device *floppy = m_floppy[m_drive]->get_device();
+			if (floppy)
+				floppy->write_flux(start_time, machine().time(), pos, buf);
+		}
+		uint8_t last = m_cmar ? m_ram[m_cmar-1] : 0;
+		m_pending_byte = mfm_byte(m_ram[m_cmar++], last);
+		m_pending_size = 16;
+		m_cmar &= 0x1ff;
+		m_byte_timer->adjust(half_bitcell_size*16);
+	}
+}
+
+void s100_vector_dualmode_device::device_start()
+{
+	m_motor_on_timer = timer_alloc(FUNC(s100_vector_dualmode_device::motor_off), this);
+	m_byte_timer = timer_alloc(FUNC(s100_vector_dualmode_device::byte_cb), this);
+	m_sector_timer = timer_alloc(FUNC(s100_vector_dualmode_device::sector_cb), this);
+
+	save_item(NAME(m_ram));
+	save_item(NAME(m_cmar));
+	save_item(NAME(m_drive));
+	save_item(NAME(m_sector));
+	save_item(NAME(m_read));
+	save_item(NAME(m_pending_byte));
+	save_item(NAME(m_pending_size));
+}
```

Comment:
Do you need to do something to register the PLL state for save states?

Reply:
I guess leave it for now if nothing else is doing it properly.  You can implement `device_post_load` if you want to get a call after a saved state is restored.